src/mlpack/methods/ann/brnn_impl.hpp

```diff
@@ -83,6 +83,15 @@ double BRNN<OutputLayerType, MergeLayerType, MergeOutputType,
     ResetParameters();
   }
 
+  if (optimizer.MaxIterations() < this->predictors.n_cols)
```

Comment:
We should check that the optimizer has a `MaxIterations()` function, so basically what you can do is:

add an entry to: https://github.com/mlpack/mlpack/blob/cfaab7877bb7498d18198127d39c3f177fe2d11b/src/mlpack/methods/ann/layer/layer_traits.hpp#L117 and than use `enable_if` to check for the function, here is an example: 

https://github.com/mlpack/mlpack/blob/537484b8cc937bcab0ed0de175626e326eac30f2/src/mlpack/methods/ann/visitor/output_width_visitor.hpp#L41

So this will not fail to build if the optimizer doesn't implement `MaxIterations()`.

Reply:
Hi @zoq, just wanted confirmation before moving forward with this. So, I have to add `HAS_MEM_FUNC()` in **layer_traits.hpp** and then use `enable_if` in `Train()` function to check if optimizer has `MaxIterations()` method in it. Am I correct?

Reply:
I would probably define another function, `CheckMaxIterations` or something that prints the warning, if `MaxIterations` is implemented, but the rest is correct.

Reply:
Right, actually it will be required to define an auxiliary function, otherwise this won't compile for optimizers that don't have a `MaxIterations()` function.  So please don't forget to add a test, so that we can know that it *will* compile in that case. :+1:

Reply:
Hey @rcurtin @zoq, sorry to disturb you again with such trivial doubts but I am stuck and require some guidance moving forward. Firstly in **layer_traits.hpp** I introduced `HAS_MEM_FUNC(MaxIterations, HasMaxIterations);` to check if the optimizer used during the training process has `MaxIterations()` member function in it and then in **ffn.hpp** I declared the function `WarnMaxIteration()` with `std::enable_if`. The code snippet was,
```
template<typename OptimizerType>
typename std::enable_if<
      HasMaxIterations<OptimizerType, size_t&(OptimizerType::*)()>::value, void>::type
void WarnMaxIteration(size_t, size_t);
```
After this in **ffn_impl.hpp** I wrote the definition of `WarnMaxIteration()`. The code snippet was,
```
template<typename OutputLayerType, typename InitializationRuleType,
         typename... CustomLayers>
template<typename OptimizerType>
inline typename std::enable_if<
    HasMaxIterations<OptimizerType, size_t&(OptimizerType::*)()>::value, size_t>::type
void FFN<OutputLayerType, InitializationRuleType, CustomLayers...>::WarnMaxIteration(
  size_t max_iterations, size_t no_of_datapoints)
{
  if(max_iterations < no_of_datapoints)
  {
    Log::Warn << "Warning Message" << std::cout;
  }
}
```
But doing this gives error. The error message says,

> two or more data types in declaration of ‘WarnMaxIteration’

Kindly help me regarding this.

Reply:
Hey @rcurtin @zoq, sorry to disturb you with such trivial doubts but I am stuck here and need your help. So according to what @zoq said first in **layer_traits.cpp** I defined `HAS_ANY_METHOD_FORM(MaxIterations, HasMaxIterations);` to check if the optimizer has `MaxIterations()` method and then in **ffn.hpp** I put the declaration of the `WarnMessage` which would print the warning message. The code snippet that I implemented is,
```
template<typename T,
typename std::enable_if<
      HasMaxIterations<T>::value, size_t>::type>
void WarnMessage(size_t no_of_iterations, size_t no_of_datapoints);
```
after this in **ffn_impl.hpp** I defined this function as,
```
template<typename T,
typename std::enable_if<
      HasMaxIterations<T>::value, size_t>::type>
void WarnMessage(size_t no_of_iterations, size_t no_of_datapoints)
{
  Log::Warn << "Warn message" << std::endl;
}
```
But when I call this function in the `Train()` function I got an error 
> no matching function for call to ‘mlpack::ann::FFN<mlpack::ann::NegativeLogLikelihood<arma::Mat<double>, arma::Mat<double> >, mlpack::ann::GlorotInitializationType<false> >::WarnMessage<ens::SGD<ens::AdamUpdate, ens::NoDecay> >(size_t&, const uword&)’


What am I doing wrong here? Kindly guide me to the right process.

Reply:
This doesn't address your error directly but don't forget that you will also need an overload with `std::disable_if<>`, to cover the case where the `MaxIterations` function does not exist.

Also, you'll need to call `MaxIterations()` *only* in the overload of the function `WarnMessage()` where `HasMaxIterations<T>::value` is `true`, otherwise we aren't guaranteed that it exists.  So probably you will need to modify the API for `WarnMessage` a bit.

I can't tell from the error message exactly what is going wrong but I'd suggest the following general steps for debugging:

 - See if the compiler is suggesting any similar alternatives to the overload it couldn't find.  Maybe there is some small mismatch.  (Could it be that `this->predictors.n_cols` has type `uword` but the API takes `size_t`?  You might need to cast it first.)

 - Check to see if `HasMaxIterations<T>::value` is the expected value, perhaps by printing that value to stdout or something.